### PR TITLE
[vectorized] [join] eliminate branch prediction & opt hash join performance

### PR DESCRIPTION
**Pull request by awakeljw**
# Proposed changes

Issue Number: close #xxx

## Problem Summary:

Opt hash join performance

1. Eliminate hashjoin probe phase branch perdiction by make_bool_variant function in compile time。
especially for innerjoin (most scenes)， left semi join left anti join(mostly used).
2. Reduce mapped memory usage，the hash map traverse mapped is time consuming，so delete the block pointer（8 Bytes）is cache friendly.

After opt, the performance is greatly improved。

My test:(SSB benchmark)
set runime_filter_type=0;
set parallel_fragment_exec_instance_num = 1;
1. inner join: 
(before opt)
MySQL [ssb1]> SELECT count(c_custkey) FROM lineorder,customer WHERE lo_custkey = c_custkey;
+--------------------+
| count(`c_custkey`) |
+--------------------+
|          600037902 |
+--------------------+
1 row in set (43.25 sec)
(after opt)
MySQL [ssb1]> SELECT count(c_custkey) FROM lineorder,customer WHERE lo_custkey = c_custkey;
+--------------------+
| count(`c_custkey`) |
+--------------------+
|          600037902 |
+--------------------+
1 row in set (30.25 sec)

2. ssb q3.3
(before opt)
MySQL [ssb1]> SELECT c_city, s_city, d_year, SUM(lo_revenue) AS  REVENUE FROM customer, lineorder, supplier, dates WHERE lo_custkey = c_custkey AND lo_suppkey = s_suppkey AND  lo_orderdate = d_datekey AND  (c_city='UNITED KI1' OR c_city='UNITED KI5') AND (s_city='UNITED KI1' OR s_city='UNITED KI5') AND d_year >= 1992 AND d_year <= 1997 GROUP BY c_city, s_city, d_year ORDER BY d_year ASC,  REVENUE DESC;
+------------+------------+--------+------------+
| c_city     | s_city     | d_year | REVENUE    |
+------------+------------+--------+------------+
| UNITED KI1 | UNITED KI1 |   1992 | 5776096629 |
| UNITED KI5 | UNITED KI1 |   1992 | 5555883901 |
| UNITED KI5 | UNITED KI5 |   1992 | 5348705805 |
| UNITED KI1 | UNITED KI5 |   1992 | 5326870427 |
| UNITED KI1 | UNITED KI1 |   1993 | 5892974670 |
| UNITED KI1 | UNITED KI5 |   1993 | 5490859451 |
| UNITED KI5 | UNITED KI1 |   1993 | 5468354303 |
| UNITED KI5 | UNITED KI5 |   1993 | 5089909647 |
| UNITED KI5 | UNITED KI1 |   1994 | 5437315108 |
| UNITED KI1 | UNITED KI1 |   1994 | 5348775917 |
| UNITED KI5 | UNITED KI5 |   1994 | 5310936695 |
| UNITED KI1 | UNITED KI5 |   1994 | 5237461110 |
| UNITED KI1 | UNITED KI1 |   1995 | 5737551920 |
| UNITED KI5 | UNITED KI5 |   1995 | 5657584590 |
| UNITED KI5 | UNITED KI1 |   1995 | 5260093556 |
| UNITED KI1 | UNITED KI5 |   1995 | 5213763257 |
| UNITED KI5 | UNITED KI1 |   1996 | 5522325005 |
| UNITED KI1 | UNITED KI1 |   1996 | 5451244409 |
| UNITED KI5 | UNITED KI5 |   1996 | 5231759057 |
| UNITED KI1 | UNITED KI5 |   1996 | 5203962897 |
| UNITED KI1 | UNITED KI1 |   1997 | 5340760807 |
| UNITED KI5 | UNITED KI1 |   1997 | 5295685214 |
| UNITED KI1 | UNITED KI5 |   1997 | 5188428156 |
| UNITED KI5 | UNITED KI5 |   1997 | 5024634475 |
+------------+------------+--------+------------+
24 rows in set (32.15 sec)
(after opt): 
MySQL [ssb1]> SELECT c_city, s_city, d_year, SUM(lo_revenue) AS  REVENUE FROM customer, lineorder, supplier, dates WHERE lo_custkey = c_custkey AND lo_suppkey = s_suppkey AND  lo_orderdate = d_datekey AND  (c_city='UNITED KI1' OR c_city='UNITED KI5') AND (s_city='UNITED KI1' OR s_city='UNITED KI5') AND d_year >= 1992 AND d_year <= 1997 GROUP BY c_city, s_city, d_year ORDER BY d_year ASC,  REVENUE DESC;
+------------+------------+--------+------------+
| c_city     | s_city     | d_year | REVENUE    |
+------------+------------+--------+------------+
| UNITED KI1 | UNITED KI1 |   1992 | 5776096629 |
| UNITED KI5 | UNITED KI1 |   1992 | 5555883901 |
| UNITED KI5 | UNITED KI5 |   1992 | 5348705805 |
| UNITED KI1 | UNITED KI5 |   1992 | 5326870427 |
| UNITED KI1 | UNITED KI1 |   1993 | 5892974670 |
| UNITED KI1 | UNITED KI5 |   1993 | 5490859451 |
| UNITED KI5 | UNITED KI1 |   1993 | 5468354303 |
| UNITED KI5 | UNITED KI5 |   1993 | 5089909647 |
| UNITED KI5 | UNITED KI1 |   1994 | 5437315108 |
| UNITED KI1 | UNITED KI1 |   1994 | 5348775917 |
| UNITED KI5 | UNITED KI5 |   1994 | 5310936695 |
| UNITED KI1 | UNITED KI5 |   1994 | 5237461110 |
| UNITED KI1 | UNITED KI1 |   1995 | 5737551920 |
| UNITED KI5 | UNITED KI5 |   1995 | 5657584590 |
| UNITED KI5 | UNITED KI1 |   1995 | 5260093556 |
| UNITED KI1 | UNITED KI5 |   1995 | 5213763257 |
| UNITED KI5 | UNITED KI1 |   1996 | 5522325005 |
| UNITED KI1 | UNITED KI1 |   1996 | 5451244409 |
| UNITED KI5 | UNITED KI5 |   1996 | 5231759057 |
| UNITED KI1 | UNITED KI5 |   1996 | 5203962897 |
| UNITED KI1 | UNITED KI1 |   1997 | 5340760807 |
| UNITED KI5 | UNITED KI1 |   1997 | 5295685214 |
| UNITED KI1 | UNITED KI5 |   1997 | 5188428156 |
| UNITED KI5 | UNITED KI5 |   1997 | 5024634475 |
+------------+------------+--------+------------+
24 rows in set (11.95 sec)
----------------------------------------------------

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (No Need)
3. Has document been added or modified: (No Need)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 1]
```diff
@@ -505,8 +604,8 @@ struct ProcessHashTableProbe {
 
 private:
     HashJoinNode* _join_node;
-    const DataTypes& _left_table_data_types;
-    const DataTypes& _right_table_data_types;
+    const int right_col_idx;
```
Comment (HappenLee) [inline be/src/vec/exec/join/vhash_join_node.cpp:607]: _right_col_idx

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 2]
```diff
@@ -26,9 +26,18 @@
 #include "vec/exprs/vexpr_context.h"
 #include "vec/functions/simple_function_factory.h"
 #include "vec/utils/util.hpp"
-
+#include <variant>
```
Comment (HappenLee) [inline be/src/vec/exec/join/vhash_join_node.cpp:29]: move 20 line and add space

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 3]
```diff
@@ -178,36 +188,133 @@ struct ProcessHashTableProbe {
...
+                }
+
+                items_counts[_probe_index++] = repeat_count;
+                current_offset += repeat_count;
+
+                if (current_offset >= _batch_size) {
+                    break;
                 }
             }
+        }
```
Comment (HappenLee) [inline be/src/vec/exec/join/vhash_join_node.cpp:245]: } else if constexpr

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 4]
```diff
@@ -178,36 +188,133 @@ struct ProcessHashTableProbe {
...
+
+                if (current_offset >= _batch_size) {
+                    break;
                 }
             }
+        } 
+        else if constexpr (is_left_anti_join) {
+            for (; _probe_index < _probe_rows;) {
+                // check ignore null
+                if constexpr (ignore_null) {
```
Comment (HappenLee) [inline be/src/vec/exec/join/vhash_join_node.cpp:249]: You need enough abstraction to use macros or lambda expressions

File: be/src/vec/core/block.h [hunk 5]
```diff
@@ -348,6 +348,13 @@ class MutableBlock {
         _data_types.clear();
     }
 
+    void append_from_block(const Block* block, const int col_len, const int start_col_idx, const size_t row_number) {
```
Comment (HappenLee) [inline be/src/vec/core/block.h:351]: the function do not call any where? should not add here

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 6]
```diff
@@ -168,46 +177,98 @@ struct ProcessHashTableProbe {
...
+                } else {
+                    auto find_result = key_getter.find_key(hash_table_ctx.hash_table, _probe_index, _arena);
+                    if (find_result.is_found()) { 
+                        //do nothing
+                    } else {
+                        ++repeat_count;
+                    }
+                }
+            } else if constexpr(is_left_semi_join) {
+                int repeat_count = 0;
```
Comment (HappenLee) [inline be/src/vec/exec/join/vhash_join_node.cpp:248]: pre define repeat count in 199line

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 7]
```diff
@@ -168,46 +177,98 @@ struct ProcessHashTableProbe {
...
+                if (!(*null_map)[_probe_index]) {
+                    auto find_result = key_getter.find_key(hash_table_ctx.hash_table, _probe_index, _arena);
+                    if (find_result.is_found()) { 
+                        ++repeat_count;
+                    }
+                }
+                items_counts[_probe_index++] = repeat_count;
```
Comment (HappenLee) [inline be/src/vec/exec/join/vhash_join_node.cpp:255]: the code have in 311line

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 8]
```diff
@@ -168,46 +177,98 @@ struct ProcessHashTableProbe {
...
                     continue;
                 }
             }
-
             int repeat_count = 0;
-            auto find_result =
-                    (*null_map)[_probe_index]
+            if constexpr (is_inner_join) {
+                if (!(*null_map)[_probe_index]) {
+                    auto find_result = key_getter.find_key(hash_table_ctx.hash_table, _probe_index, _arena);
```
Comment (HappenLee) [inline be/src/vec/exec/join/vhash_join_node.cpp:202]: why not pretch here ?

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 9]
```diff
@@ -168,46 +177,98 @@ struct ProcessHashTableProbe {
...
+                    ++repeat_count;
+                } else {
+                    auto find_result = key_getter.find_key(hash_table_ctx.hash_table, _probe_index, _arena);
+                    if (find_result.is_found()) { 
+                        //do nothing
+                    } else {
+                        ++repeat_count;
+                    }
+                }
+            } else if constexpr(is_left_semi_join) {
```
Comment (HappenLee) [inline be/src/vec/exec/join/vhash_join_node.cpp:247]: why not contexpr more join type, like right semi/anti join ?

**Review by HappenLee — COMMENTED** [review]
some little problem

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 10]
```diff
@@ -711,45 +766,62 @@ Status HashJoinNode::get_next(RuntimeState* state, Block* output_block, bool* eo
...
     if (_probe_index < _probe_block.rows()) {
-        MutableBlock mutable_block = (output_block->mem_reuse() && !_have_other_join_conjunct) ?
-                MutableBlock(output_block) :
-                MutableBlock(VectorizedUtils::create_empty_columnswithtypename(
-                        !_have_other_join_conjunct ? row_desc() : _row_desc_for_other_join_conjunt));
-
         std::visit(
-                [&](auto&& arg) {
+            [&](auto&& arg, auto have_other_join_conjunct, 
+            auto is_left_semi_join, auto is_left_anti_join, auto is_inner_join) {
```
Comment (HappenLee) [inline be/src/vec/exec/join/vhash_join_node.cpp:771]: why keep so many bool, maybe you should use std::variant<std::monostate, Inner Join, Semi Join, Anti Join

File: be/src/vec/exec/join/vhash_join_node.cpp (hunk 8 above)
Comment (awakeljw) [reply be/src/vec/exec/join/vhash_join_node.cpp:202]: in my test ,wo found prefetch is time consuming，but not helpful for performance ,so maybe when mapped is more than 1, it is helpful.

File: be/src/vec/exec/join/vhash_join_node.cpp (hunk 9 above)
Comment (awakeljw) [reply be/src/vec/exec/join/vhash_join_node.cpp:247]: left semi , left anti, inner join is mostly used.so i just eliminate the three branch. others join will benefit from it

File: be/src/vec/exec/join/vhash_join_node.cpp (hunk 10 above)
Comment (awakeljw) [reply be/src/vec/exec/join/vhash_join_node.cpp:771]: these three parameter is std::integral_constant, not bool. use std::variant, we also need to distinguish it in next execution. and i think std::variant is not efficient.

File: be/src/vec/core/block.h (hunk 5 above)
Comment (awakeljw) [reply be/src/vec/core/block.h:351]: i agree，i will delete it.

File: be/src/vec/exec/join/vhash_join_node.cpp (hunk 7 above)
Comment (awakeljw) [reply be/src/vec/exec/join/vhash_join_node.cpp]: i agree, and has delete it.

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 11]
```diff
@@ -804,6 +838,10 @@ Status HashJoinNode::_hash_table_build(RuntimeState* state) {
     RETURN_IF_ERROR(child(1)->open(state));
     SCOPED_TIMER(_build_timer);
     Block block;
+    MutableColumns columns(_right_table_data_types.size());
+    for (int i = 0; i < _right_table_data_types.size(); i++) {
+        columns[i] = _right_table_data_types[i]->create_column();
```
Comment (zuochunwei) [inline be/src/vec/exec/join/vhash_join_node.cpp:855]: call reserve() here to avoid exlarging multi-times?

File: be/src/vec/exec/join/join_op.h [hunk 12]
```diff
@@ -26,15 +26,15 @@ namespace doris::vectorized {
 struct RowRef {
     using SizeT = uint32_t; /// Do not use size_t cause of memory economy
 
-    const Block* block = nullptr;
+    // const Block* block = nullptr;
     SizeT row_num = 0;
     // Use in right join to mark row is visited
     // TODO: opt the varaible to use it only need
     bool visited = false;
```
Comment (zuochunwei) [inline be/src/vec/exec/join/join_op.h:33]: use bitfield to make visited as the highest bit of row_num? maybe useful.

File: be/src/vec/exec/join/join_op.h [hunk 13]
```diff
@@ -115,7 +115,7 @@ struct RowRefList : RowRef {
     };
 
     RowRefList() {}
-    RowRefList(const Block* block_, size_t row_num_) : RowRef(block_, row_num_) {}
+    RowRefList(size_t row_num_) : RowRef(row_num_) {}
 
     ForwardIterator begin() { return ForwardIterator(this); }
```
Comment (zuochunwei) [inline be/src/vec/exec/join/join_op.h:120]: iterating RowRefList through ForwardIterator is slow, consider providing a interface to get all row_num list

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 14]
```diff
@@ -177,37 +180,66 @@ struct ProcessHashTableProbe {
...
+
+                        if (mapped.get_row_count() == 1) {
+                            ++repeat_count;
+                            _build_index.emplace_back(mapped.row_num);
+                        } else {
+                            if (_probe_index + 2 < _probe_rows)
+                                key_getter.prefetch(hash_table_ctx.hash_table, _probe_index + 2, _arena);
+                            for (auto it = mapped.begin(); it.ok(); ++it) {
+                                ++repeat_count;
+                                _build_index.emplace_back(it->row_num);
```
Comment (zuochunwei) [inline be/src/vec/exec/join/vhash_join_node.cpp:210]: if row_num stored in mapped as continuous array, we can just remember the row_num array’s address here. 
it's no need to copy one by one.

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 15]
```diff
@@ -292,8 +318,9 @@ struct ProcessHashTableProbe {
         std::vector<bool> same_to_prev;
         same_to_prev.reserve(1.2 * _batch_size);
 
-        int right_col_idx = _left_table_data_types.size();
-        int right_col_len = _right_table_data_types.size();
+        std::vector<uint32_t> _build_index;
+        _build_index.reserve(1.2 * _batch_size);
```
Comment (zuochunwei) [inline be/src/vec/exec/join/vhash_join_node.cpp:322]: consider using stack array instead?

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 16]
```diff
@@ -316,10 +343,7 @@ struct ProcessHashTableProbe {
 
                 for (auto it = mapped.begin(); it.ok(); ++it) {
                     ++current_offset;
-                    for (size_t j = 0; j < right_col_len; ++j) {
-                        auto& column = *it->block->get_by_position(j).column;
-                        mcol[j + right_col_idx]->insert_from(column, it->row_num);
-                    }
+                    _build_index.emplace_back(it->row_num);
```
Comment (zuochunwei) [inline be/src/vec/exec/join/vhash_join_node.cpp:346]: increase current_offset += it.size() out of the loop

File: be/src/vec/exec/join/join_op.h [hunk 17]
```diff
@@ -26,15 +26,15 @@ namespace doris::vectorized {
...
-    const Block* block = nullptr;
+    // const Block* block = nullptr;
     SizeT row_num = 0;
     // Use in right join to mark row is visited
     // TODO: opt the varaible to use it only need
     bool visited = false;
 
     RowRef() {}
-    RowRef(const Block* block_ptr, size_t row_num_count, bool is_visited = false)
-            : block(block_ptr), row_num(row_num_count), visited(is_visited) {}
```
Comment (zuochunwei) [inline be/src/vec/exec/join/join_op.h:37]: i am not sure whether placing member data 'row_count' before 'next' will save memory, try it.

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 18]
```diff
@@ -177,89 +180,113 @@ struct ProcessHashTableProbe {
...
+                        auto& mapped = find_result.get_mapped();
+
+                        if (mapped.get_row_count() == 1) {
+                            ++repeat_count;
+                            _build_index.emplace_back(mapped.row_num);
+                        } else {
+                            if (_probe_index + 2 < _probe_rows)
+                                key_getter.prefetch(hash_table_ctx.hash_table, _probe_index + 2, _arena);
+                            for (auto it = mapped.begin(); it.ok(); ++it) {
+                                ++repeat_count;
```
Comment (zuochunwei) [inline be/src/vec/exec/join/vhash_join_node.cpp:209]: increase repeat_count value out of the loop

File: be/src/vec/exec/join/vhash_join_node.cpp [hunk 19]
```diff
@@ -217,49 +249,44 @@ struct ProcessHashTableProbe {
...
                 break;
             }
         }
-        
-        for (int i = 0; i < right_col_idx; ++i) {
+
+        // insert all match build rows
+        for (int i = 0; i < _right_col_len; i++) {
+            auto &column = *_build_block.get_by_position(i).column;
+            for (int j = 0; j < _build_index.size(); j++) {
```
[the remark this answers is not in the document]
Comment (zuochunwei) [reply be/src/vec/exec/join/vhash_join_node.cpp:284]: avoid call the virtual function insert_from in loop

**Pull request closed without merging**